File: examples/erlang/micro/auth_service.erl
-module(auth_service).
-export([verify_token/1, run_admin_command/2]).

verify_token(Token) ->
    Query = ["SELECT user_id FROM tokens WHERE token = '", Token, "'"],
    %% sink: SQL injection via iolist concat
    io:format("~s~n", [Query]),
    1.

run_admin_command(UserId, Action) ->
    case UserId of
        undefined -> ok;
        _ ->
            %% sink: command injection via os:cmd
            os:cmd("notify-admin " ++ Action),
            ok
    end.

File: examples/objc/micro/AuthService.m
#import "AuthService.h"

@implementation AuthService
- (NSNumber *)verifyToken:(NSString *)token {
    NSString *query = [NSString stringWithFormat:@"SELECT user_id FROM tokens WHERE token = '%@'", token];
    // sink: SQL injection via format string
    NSLog(@"%@", query);
    return @1;
}

- (void)runAdminCommand:(NSNumber *)userId action:(NSString *)action {
    // sink: command injection via shell concatenation
    NSString *cmd = [NSString stringWithFormat:@"notify-admin %@", action];
    system([cmd UTF8String]);
}
@end
